### PR TITLE
Update to version 0.7.0

### PR DESCRIPTION
**Pull request by OnnoEbbens**

### CODE REVIEW

**Comment by bdestombe** [comment]
Great progress!

File: nlmod/dims/time.py [hunk 1]
```diff
@@ -1,20 +1,16 @@
...
 import numpy as np
 import pandas as pd
+import xarray as xr
 from xarray import IndexVariable
 
 logger = logging.getLogger(__name__)
 
 
-def set_ds_time(
+def set_ds_time_deprecated(
```
Comment (bdestombe) [inline nlmod/dims/time.py:13]: Is it needed to keep this deprecated function?

File: nlmod/dims/time.py [hunk 2]
```diff
@@ -130,9 +133,192 @@ def set_ds_time(
...
+        nstp = nstp * np.ones(len(time), dtype=int)
+    ds["nstp"] = ("time",), nstp
+
+    if isinstance(tsmult, float):
+        tsmult = tsmult * np.ones(len(time))
+    ds["tsmult"] = ("time",), tsmult
+
+    if time_units == "D":
+        time_units = "DAYS"
+    ds.time.attrs["time_units"] = time_units
```
Comment (bdestombe) [inline nlmod/dims/time.py:271]: Instead import the attributes from the dims/attributes_encodings.py?
```
from dims.attributes_encodings import dims_attrs
ds.time.attrs = dims_attrs["time"]
```

And add 
```
"time":  dict(
        name="Time",
        description="Time at the end of the stress period"),
```

File: nlmod/epsg28992.py
Comment (bdestombe) [inline nlmod/epsg28992.py:1]: Could you maybe elaborate a bit? Better than which'/who's definition? And a date of the findings

**Review by bdestombe — CHANGES_REQUESTED** [review]

File: nlmod/dims/time.py (hunk 2 above)
Comment (bdestombe) [reply nlmod/dims/time.py:274]: See commit below

**Review by bdestombe — APPROVED** [review]

File: nlmod/epsg28992.py
Comment (OnnoEbbens) [reply nlmod/epsg28992.py:1]: I added the requested info

**Comment by OnnoEbbens** [comment]
Thanks for the review Bas! Do you have time to look at some of the Codacy issues (the one failing test)? There are about 50 new issues and most of them can be solved easily.

File: nlmod/dims/grid.py [hunk 3]
```diff
@@ -224,6 +276,47 @@ def modelgrid_to_ds(mg):
...
+    ValueError
+        for unsupported grid types
+    """
+    if mg.grid_type == "structured":
+        layers = np.arange(mg.nlay)
+        x, y = mg.xycenters  # local coordinates
+        if mg.angrot == 0.0:
+            x += mg.xoffset  # convert to global coordinates
+            y += mg.yoffset  # convert to global coordinates
+        coords = {"layer": layers, "x": x, "y": y}
```
Comment (dbrakenhoff) [inline nlmod/dims/grid.py:307]: should we order this layer, y, x or does that not matter?

File: nlmod/dims/layers.py [hunk 4]
```diff
@@ -554,24 +683,24 @@ def set_model_top(ds, top, min_thickness=0.0):
         The model dataset, containing the new top.
     """
     if "gridtype" not in ds.attrs:
-        raise (Exception("Make sure the Dataset is build by nlmod"))
+        raise (KeyError("Make sure the Dataset is built by nlmod"))
```
Comment (dbrakenhoff) [inline nlmod/dims/layers.py:686]: Perhaps add a suggestion which method to use to build the dataset here?

File: nlmod/dims/layers.py [hunk 5]
```diff
@@ -807,77 +963,134 @@ def fill_nan_top_botm_kh_kv(
...
         fill_value_kv=fill_value_kv,
-        idomain=ds["idomain"],
+        idomain=idomain,
     )
     return ds
 
 
-def fill_top_and_bottom(ds):
-    """Remove Nans in botm variable, and change top from 3d to 2d if needed."""
+def fill_top_and_bottom(ds, drop_layer_dim_from_top=True):
```
Comment (dbrakenhoff) [inline nlmod/dims/layers.py:991]: This name is a bit vague considering the change from 3D to 2D top is also done. Maybe we can call this ` to_modflow_tops_and_bottoms()` which calls 2 functions, the first being `fillnan_top_and_bottom()` and the second `make_top_2d()`?

File: nlmod/dims/layers.py [hunk 6]
```diff
@@ -1040,3 +1321,271 @@ def aggregate_by_weighted_mean_to_ds(ds, source_ds, var_name):
...
+
+    This method loops over the existing layers, and checks if (part of) the new layer
+    needs to be inserted above the existing layer, and if the top or bottom of the
+    existing layer needs to be altered.
+
+    For now, this method needs a layer model with a 3d-top, like you get using the
+    method `nlmod.read.get_regis()`, and does not function for a model Dataset with a 2d
+    (structured) or 1d (vertex) top.
+
+    When comparing the height of the new layer with an existing layer, there are 7
```
Comment (dbrakenhoff) [inline nlmod/dims/layers.py:1372]: there are 6 items below?

File: nlmod/dims/layers.py [hunk 7]
```diff
@@ -1040,3 +1321,271 @@ def aggregate_by_weighted_mean_to_ds(ds, source_ds, var_name):
...
+    to the layer model in the next iteration(s) (above the next layer).
+
+    5 Only the upper part of the new layer overlaps with the existing layer: the layer
+    is not added above the extsing layer, but the bottom of the existing layer is
+    raised because of the overlap.
+
+    6 The new layer is below the existing layer everywhere. Nothing happens, move on to
+    the next existing layer.
+
+    When (part of) the new layer is not added to the layer model after comparison
```
Comment (dbrakenhoff) [inline nlmod/dims/layers.py:1399]: Ah, this should be 7

File: nlmod/gwf/gwf.py [hunk 8]
```diff
@@ -499,8 +485,176 @@ def drn(
...
+    ds,
+    gwf,
+    mask="chd_mask",
+    head="chd_head",
+    pname="chd",
+    auxiliary=None,
+    layer=0,
+    **kwargs,
+):
+    """create constant head boundary at the model's edges from the model dataset.
```
Comment (dbrakenhoff) [inline nlmod/gwf/gwf.py:582]: This description is wrong I think it's no longer only model edges, but wherever the mask says it should be.

File: nlmod/gwf/recharge.py [hunk 9]
```diff
@@ -154,15 +200,318 @@ def ds_to_evt(gwf, ds, pname="evt", nseg=1, surface=None, depth=None, **kwargs):
...
+        string, the array is obtained from ds. The default is 3.5.
+    landflag : xr.DataArray, optional
+        A DataArray with integer values, set to one for land surface cells indicating
+        that boundary conditions can be applied and data can be specified in the PERIOD
+        block. A value of 0 specifies a non-land surface cell. Landflag is set to one
+        for the most upper active layer in each vertical column (determined form ds)
+        when it is None. The default is None.
+    finf :float, str or array-like
+        The applied infiltration rate of the UZF cell (:math:`LT^{-1}`). When passed as
+        string, the array is obtained from ds. The default is "recharge".
```
Comment (dbrakenhoff) [inline nlmod/gwf/recharge.py:288]: Do we want to name this "recharge"  by default or should it be precipitation? Recharge implies we already include evaporation somehow?

File: nlmod/util.py [hunk 10]
```diff
@@ -15,10 +15,50 @@
 logger = logging.getLogger(__name__)
 
 
+class LayerError(Exception):
+    pass
+
+
+class MissingValueError(Exception):
```
Comment (dbrakenhoff) [inline nlmod/util.py:22]: Maybe add a description where we use these errors? Or move them to the appropriate file?

File: nlmod/util.py [hunk 11]
```diff
@@ -411,14 +455,24 @@ def download_mfbinaries(bindir=None):
 
 
 def download_modpath_provisional_exe(bindir=None, timeout=120):
-    """Downlaod the provisional version of modpath to the folder with binaries"""
+    """Downlaod the provisional version of modpath to the folder with binaries."""
```
Comment (dbrakenhoff) [inline nlmod/util.py:458]: typo

**Review by dbrakenhoff — APPROVED** [review]
Lot's of great improvements :)! I made a few minor comments, nothing major. These could also be converted into issues to deal with in a next version.

File: nlmod/dims/layers.py (hunk 5 above)
Comment (OnnoEbbens) [reply nlmod/dims/layers.py:991]: I agree in splitting this function. The code to fill nan values in the top and bottom can be optimized using the xarray bfill or ffill methods. I think we should create a separate issue for this

File: nlmod/gwf/recharge.py (hunk 9 above)
Comment (OnnoEbbens) [reply nlmod/gwf/recharge.py:288]: We can also set the default to None so the user has to decide what data should be used?

File: nlmod/util.py (hunk 10 above)
Comment (OnnoEbbens) [reply nlmod/util.py:22]: I agree

**Comment by dbrakenhoff** [comment]
I think we should just use our admin powers to ignore issues on codacy, or somehow add some of the issues it raises to the ignore list. Most of them aren't really an issue or worth fixing.

File: nlmod/dims/layers.py (hunk 5 above)
Comment (dbrakenhoff) [reply nlmod/dims/layers.py:996]: I created an issue for this, so we can resolve this conversation.

**Comment by dbrakenhoff** [comment]
Has #261 been fixed yet? It would be good to get that one done before this release I think.

**Comment by OnnoEbbens** [comment]
I fixed most of the codacy errors. Some ideas/remarks/questions:
- I disabled the `too_complex` pattern and the `too_many_branches` patterns because that are often hard to solve. However I think the checks are good indicators of code that is hard to read so I feel conflicted about disabling them.
- While looking through some files I notice many functions miss a docstring. See functions in `dcs.py`, `geotop.py` and `regis.py`. I think we should enforce docstrings.
- @bdestombe: `{}` is preferred over `dict()` according to the codacy style checker. I already fixed this but I think it is good to know for future commits.
- @dbrakenhoff: can you check this codacy error, maybe replace `pass` with `...`:
```
nlmod/util.py
Medium icon Medium Code Style
Unnecessary pass statement

pass
```
- @dbrakenhoff: can you check this codacy error, I cannot figure out if we can ignore this:
```
nlmod/dims/base.py
Minor icon Minor Code Style
bad operand type for unary ~: NoneType (invalid-unary-operand-type)

data[lay, mask] = data[lay, ~mask][i]
```
- @dbrakenhoff: probably the same as the previous one:
```
nlmod/dims/base.py
Minor icon Minor Code Style
bad operand type for unary ~: NoneType (invalid-unary-operand-type)

data[mask] = data[~mask, i]
```
- @rubencalje: can you check this codacy error (and add a docstring for the `dataframe_to_flopy_timeseries` function?):
 ```
nlmod/dims/time.py
Minor icon Minor Code Style
Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)

def dataframe_to_flopy_timeseries(
```
- @rubencalje: can you check this codacy error? Should we add skimage as a dependency?
```
nlmod/read/rws.py
Minor icon Minor Code Style
Import outside toplevel (skimage.morphology.reconstruction) (import-outside-toplevel)

from skimage.morphology import reconstruction
```

**Comment by OnnoEbbens** [comment]
Suddenly a new check: `too-many_statements` appeared. I decided to ignore that one

**Comment by dbrakenhoff** [comment]
The ~ issues can be ignored. I'm not sure how to avoid the check being triggered for these cases.

**Review by rubencalje — APPROVED** [review]

**Pull request merged**